include/nanobind/nb_class.h

```diff
@@ -340,6 +340,16 @@ class class_ : public object {
         sizeof...(Ts) == !std::is_same_v<Base, T> + !std::is_same_v<Alias, T>,
         "nanobind::class_<> was invoked with extra arguments that could not be handled");
 
+    // Fail on virtual bases -- they need a this-ptr adjustment, but they're
+    // not amenable to the runtime test in the class_ constructor (because
+    // a C-style cast will do reinterpret_cast if static_cast is invalid).
```

Comment:
I don't think we should go out of our way to support non-public bases. In fact, the intention of such an interface is precisely to avoid the sort of `static_cast` that nanobind would otherwise perform when passing `Derived` to functions expecting `Base`.

Reply:
Fair enough! That should allow removing the virtual-base-specific checks.